### PR TITLE
Adds custom lighting support to modular computers, and a flashlight to tablets

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
add: Tablets now have a built-in flashlight! It can even change colors, as long as they're light enough in hue.
/:cl:

The biggest reason I hear people disliking tablets for is that the PDA has a flashlight, which makes it superior in their eyes. No longer! Tablets now have a built-in flashlight, equal in brightness to the PDA's. Unlike the PDA's, however, the tablet's flashlight can be changed in color! It does, however, only support fairly bright colors (R, G, and B values must be greater than 200 to be accepted.)

This also adds support for modular computers in general to emit dynamic light of variable range and color, although this is currently only used in tablets. To-do: Move the light itself to a hardware module for the computer itself - I just wanted to get the framework done and PR made to get thoughts before I took the time to dive into the hardware code.

### CODE REVIEW

**Comment by KorPhaeron** [comment]
Why do we have tablets + pdas if tablets do everything pdas do but more

File: code/modules/modular_computers/computers/item/computer.dm [hunk 1]
```diff
@@ -43,6 +43,10 @@
 
 	var/list/idle_threads							// Idle programs on background. They still receive process calls but can't be interacted with.
 	var/obj/physical = null									// Object that represents our computer. It's used for Adjacent() and UI visibility checks.
+	var/has_light = FALSE						//If the computer has a flashlight/LED light/what-have-you installed
+	var/light_on = FALSE						//If that light is enabled
+	var/comp_light_luminosity = 3				//The brightness of that light
+	var/comp_light_color = "#FFFFFF"			//The color of that light
```
Comment (Cyberboss) [inline code/modules/modular_computers/computers/item/computer.dm]: Have this be null by default, it's easier on the lighting system

**Comment by Xhuis** [comment]
I can't give an unbiased answer to that question, unfortunately. As it stands, I'm trying to give an incentive to use them, as they have a niche that they seem designed to fill but are neglected due to lack of stuff like a flashlight.

**Review by Cyberboss — DISMISSED** [review]

File: code/modules/modular_computers/computers/item/computer.dm (hunk 1 above)
Comment (Xhuis) [inline code/modules/modular_computers/computers/item/computer.dm]: Would it turn to #FFFFFF in `Initialize()`, or?

**Comment by GunHog** [comment]
Thank you, cat-slime-Xhuis! You are doing great work!

> Why do we have tablets + pdas if tablets do everything pdas do but more

This is not yet the case, Kor. Tablets are missing most PDA cart programs and other functions like a good messenger (the chat client is not good enough to replace messaging).

**Comment by Xhuis** [comment]
The chat client also fills a different niche. For instance, allow me to give you an idea next time you're a traitor: make a chat channel, with the password as all of the code words, lowercase, with no spaces, and include that the password is the code words in the title. Any traitor can access it immediately and for free, but anyone else has no access!

**Comment by Shadowlight213** [comment]
Bay recently made an email program. Ive been thinking of porting that, but I havent looked into it yet, and it may be too complex for players. Bays version has shit like email addresses and passwords.

**Comment by Tacolizard** [comment]
@Shadowlight213 even with this PR merged, nobody will use consoles. People barely even use PDAs. IMO we should just port all PDA features to consoles and remove PDAs.

**Comment by GunHog** [comment]
In order for tablets to be a replacement for PDAs without loss of functionality:
- Tablet should start on, for the purpose of messaging/hacking/bomb virus
- Tablet should have a messenger/email program that works by name (email address is fine if it has the person's name, e.g. Magni.Bronzebeard@NTnet.nt). The program should also post to the user's chat box and allow replies from it, just as PDAs do now.
- Tablets should have a matching program for each PDA cart program existing today.
- Tablets must work as an ID for all cases that PDAs do (tablets do not work for some things that require a manual swipe)
- All crew should start with one.

> People barely even use PDAs

I use my PDA every single round, as most other players use the PDA for lighting, messaging, and holding IDs. If you mean _tablets_, then this would be a more understandable statement.

**Comment by YakumoChen** [comment]
>People barely even use PDAs.

I disagree, I used my PDA all the time, it's especially useful with all the nifty things the carts provide. The gas scanner function is valuable for toxins and atmos, and works as an analyzer I don't need in my pocket. It's also great for food testing if you have the reagent scanner. The bot call is helpful in a pinch since I can summon a medibot if I think I'm done for and won't get medical help. That's not even to start with the fact it can hold a pen (valuable for someone who likes enforcing paperwork) and my ID and is a free flashlight everyone starts with. Hell, I never use the notetaking function but I imagine it'd also be useful for doing genetics or something.

**Comment by ike709** [comment]
I actually message people using my PDA on a regular basis. The only reason I don't use a tablet instead is because it lacks features like a flashlight and a decent messaging program, and this PR fixes one of those issues.

**Comment by Tacolizard** [comment]
Ok I revise my statement: people barely use PDAs except for the flashlight and messaging. And occasionally gas scanning.

**Comment by GunHog** [comment]
> Ok I revise my statement: people barely use PDAs except for the flashlight and messaging. And occasionally gas scanning.

I also use mine for holding my ID, my pAI, my pen-disguised antag weapons, calling bots when I need one, and remotely triggering bombs. If I have a health scanner in my PDA (say I am playing a medical role), I will use it instead of a dedicated handheld.

I am sorry, but "people barely use PDAs" does not seem accurate from my personal observations. It is one of the most important and useful items a crewmember has, second to his ID (which for many is inside the PDA).

**Comment by Nabski89** [comment]
I used to use the notekeeping function for genetics Yakumo, but the addnotes feature is a little easier, since it's a chat command and doesn't require navigating off the main PDA screen. That being said since we're loving on PDAs. The health scanner and gas scanner are my go too uses. I'll break in and steal a lawyer card every now and then for fun.

**Comment by Bawhoppen** [comment]
Make PDAs fit in wallets for maximum asymmetrical balance

**Comment by octareenroon91** [comment]
PDAs also have the not-tgui speed advantage.

**Comment by Jalleo** [comment]
@Shadowlight213 you can replace the password system by using IDs and giving each player a specific ID (letter and number password) to change their email to a new ID when required.

**Comment by Jalleo** [comment]
Or make it a part of the HoP role to move accounts around in regards of the emails.

**Comment by Bawhoppen** [comment]
More HoP stuff is good do it

**Review by Cyberboss — APPROVED** [review]
Tablets > PDAs

**Comment by Xhuis** [comment]
Tablets aren't greater than PDAs! They're equal in value.

**Pull request merged**